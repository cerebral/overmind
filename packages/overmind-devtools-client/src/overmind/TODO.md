- We should queue mutations and pop them off in our state observer
- Show action arguments
-
