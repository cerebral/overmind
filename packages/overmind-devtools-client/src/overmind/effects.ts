import BackendConnector from '../BackendConnector'

export const connector = new BackendConnector()

export const utils = {
  confirmDialog(text): boolean {
    return window.confirm(text)
  },
  getPort() {
    return Number(
      window.__OVERMIND_DEVTOOLS_BACKEND_PORT__ ||
        location.search.split('=')[1] ||
        3031
    )
  },
}

export const storage = {
  clear() {
    return new Promise<void>((resolve, reject) => {
      connector.send('storage:clear', null, (error) => {
        if (error) {
          reject(error)
        } else {
          resolve()
        }
      })
    })
  },
  get<T>(key: string): Promise<T> {
    return new Promise((resolve, reject) => {
      connector.send('storage:get', { key }, (err, data) => {
        if (err) {
          reject(err)
        } else {
          resolve(data)
        }
      })
    })
  },
  set(key, data) {
    return new Promise<void>((resolve, reject) => {
      connector.send(
        'storage:set',
        {
          key,
          data,
        },
        (err) => {
          if (err) {
            reject(err)
          } else {
            resolve()
          }
        }
      )
    })
  },
}
