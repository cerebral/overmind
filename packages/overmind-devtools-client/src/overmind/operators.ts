import { Context } from 'vm'

import {
  ActionGroupItem,
  ActionItem,
  ActionsListItemType,
  AddComponentMessage,
  AppMessage,
  ChartMessage,
  DerivedMessage,
  DirtyDerivedMessage,
  Effect,
  EffectMessage,
  EndActionMessage,
  EndOperatorMessage,
  EventType,
  ExecutionType,
  FlushMessage,
  GetterMessage,
  InitMessage,
  Message,
  MutationsMessage,
  RemoveComponentMessage,
  StartActionMessage,
  StartOperatorMessage,
  StateMessage,
  UpdateComponentMessage,
} from './types'
import {
  createApp,
  ensureFlushExists,
  getActionId,
  getOperatorId,
  runMutation,
} from './utils'

export const ensureCurrentApp = ({ state }: Context, message: Message) => {
  if (!state.currentAppName) {
    state.currentAppName = message.appName
  }

  return message
}

export const runGetterMutation = ({ state }, message: GetterMessage) => {
  runMutation(state.apps[message.appName].state)({
    method: 'set',
    path: message.data.path,
    args: [message.data.value],
    delimiter: state.apps[message.appName].delimiter || '.',
  })
}

export const addStateAndActions = ({ state }, message: InitMessage) => {
  state.isConnecting = false
  state.error = null
  state.apps[message.appName].connectionState = 'connected'
  state.apps[message.appName].state = message.data.state
  state.apps[message.appName].actionPaths = message.data.actions
  state.apps[message.appName].delimiter = message.data.delimiter || '.'
}

export const addFlush = ({ state }: Context, message: FlushMessage) => {
  ensureFlushExists(state.apps[message.appName].flushes, message.data)
  state.apps[message.appName].flushes[message.data.flushId].mutations =
    message.data.mutations

  if (typeof message.data.operatorId === 'undefined') {
    state.apps[message.appName].flushByActionId[getActionId(message.data)] = {
      flushId: message.data.flushId,
    }
  } else {
    state.apps[message.appName].flushByOperatorId[getOperatorId(message.data)] =
      {
        flushId: message.data.flushId,
      }
  }
}

export const ensureApp = ({ state, actions }: Context, message: Message) => {
  if (
    !state.apps[message.appName] ||
    message.message.type === ExecutionType.INIT
  ) {
    state.apps[message.appName] = createApp({
      name: message.appName,
    })

    // Do not use await, as it will block further execution, this stuff has to be sync
    actions.setAppDataFromStorage({
      appName: message.appName,
      actions: message.message.data.actions,
    })
  }

  return message
}

export const addClientMessage = ({ state }: Context, message: Message) => {
  state.apps[message.appName].messages.unshift(message.message)

  return message
}

export const addComponent = (
  { state }: Context,
  message: AddComponentMessage
) => {
  const id = `${message.data.componentId}_${message.data.componentInstanceId}`

  state.apps[message.appName].components[id] = {
    id,
    isMounted: true,
    updateCount: 0,
    name: message.data.name,
    paths: message.data.paths,
    isCollapsed: true,
  }
}

export const updateComponent = (
  { state }: Context,
  message: UpdateComponentMessage
) => {
  const id = `${message.data.componentId}_${message.data.componentInstanceId}`

  state.apps[message.appName].components[id].paths = message.data.paths

  if ('flushId' in message.data) {
    state.apps[message.appName].components[id].updateCount++

    ensureFlushExists(state.apps[message.appName].flushes, message.data)

    if (
      state.apps[message.appName].flushes[
        message.data.flushId
      ].components.includes(id)
    ) {
      return
    }

    state.apps[message.appName].flushes[message.data.flushId].components.push(
      id
    )
  }
}

export const removeComponent = (
  { state }: Context,
  message: RemoveComponentMessage
) => {
  const id = `${message.data.componentId}_${message.data.componentInstanceId}`

  state.apps[message.appName].components[id].isMounted = false
}

export const updateDerived = ({ state }: Context, message: DerivedMessage) => {
  const appState = state.apps[message.appName].state
  // For backwards compatability it can still be a string
  const actualPath = Array.isArray(message.data.path)
    ? message.data.path
    : (message.data.path as any).split('.')
  const path = actualPath.slice()
  const key = path.pop()
  const target = path.reduce(
    (aggr, pathKey) =>
      aggr && aggr.__CLASS__ ? aggr[pathKey].value : aggr[pathKey],
    appState
  )

  if (target.__CLASS__) {
    target.value[key] = message.data.value
  } else {
    target[key] = message.data.value
  }

  state.apps[message.appName].derived[
    actualPath.join(state.apps[message.appName].delimiter)
  ] = message.data
}

export const updateFlushWithDerived = (
  { state }: Context,
  message: DirtyDerivedMessage
) => {
  ensureFlushExists(state.apps[message.appName].flushes, message.data)
  state.apps[message.appName].flushes[message.data.flushId].derived.push(
    message.data.derivedPath
  )
}

export const addAction = ({ state }: Context, message: StartActionMessage) => {
  const app = state.apps[message.appName]
  const action = message.data
  const actionId = getActionId(action)
  const isSelectingFirstAction =
    !app.currentActionId || app.currentActionId === app.actionsList[0].id

  app.actions[actionId] = {
    ...action,
    isRunning: true,
    operators: {},
    hasError: false,
    time: Date.now(),
  }

  if (action.parentExecution) {
    return
  }

  if (
    !app.actionsList.length ||
    app.actionsList[0].actionId !== action.actionId
  ) {
    app.actionsList.unshift({
      type: ActionsListItemType.ACTION,
      id: actionId,
      actionId: action.actionId,
    })
  } else if (app.actionsList[0].type === ActionsListItemType.ACTION) {
    const existingId = (app.actionsList[0] as ActionItem).id
    app.actionsList[0] = {
      type: ActionsListItemType.GROUP,
      id: actionId,
      actionId: action.actionId,
      isCollapsed: true,
      actionIds: [actionId, existingId],
    }
  } else if (app.actionsList[0].type === ActionsListItemType.GROUP) {
    ;(app.actionsList[0] as ActionGroupItem).actionIds.unshift(actionId)
    ;(app.actionsList[0] as ActionGroupItem).id = actionId
  }

  if (isSelectingFirstAction) {
    app.currentActionId = actionId
  }
}

export const addChart = ({ state }: Context, message: ChartMessage) => {
  const app = state.apps[message.appName]
  const data = message.data
  const chartId = data.path.join('.')
  const isSelectingFirstChart = !app.currentChartId

  app.charts[chartId] = data

  if (isSelectingFirstChart) {
    app.currentChartId = chartId
  }
}

export const addOperator = (
  { state }: Context,
  message: StartOperatorMessage
) => {
  const operatorData = message.data
  const actionId = getActionId(operatorData)
  const action = state.apps[message.appName].actions[actionId]
  const operator = {
    ...operatorData,
    isRunning: true,
    events: [],
  }

  if (operatorData.parentExecution && operatorData.operatorId === 0) {
    const parentActionId = getActionId(operatorData.parentExecution)
    const parentOperator =
      state.apps[message.appName].actions[parentActionId].operators[
        operatorData.parentExecution.operatorId
      ]

    parentOperator.events.push({
      type: EventType.Action,
      data: actionId,
    })
  }

  action.operators[operator.operatorId] = operator
}

export const updateOperator = (
  { state }: Context,
  message: EndOperatorMessage
) => {
  const operatorData = message.data
  const actionId = getActionId(operatorData)
  const action = state.apps[message.appName].actions[actionId]
  const operator = action.operators[operatorData.operatorId]

  operator.isAsync = operatorData.isAsync
  operator.isRunning = false
  operator.result = operatorData.result
  operator.error = operatorData.error
  operator.isIntercepted = operatorData.isIntercepted
  operator.isSkipped = operatorData.isSkipped

  if (operator.error) {
    action.hasError = true
  }
}

export const updateState = ({ state }: Context, message: StateMessage) => {
  const app = state.apps[message.appName]
  const path = message.data.path
  const key = path.pop()
  const target = path.reduce(
    (aggr, key) => (aggr[key].__CLASS__ ? aggr[key].value : aggr[key]),
    app.state
  )

  target[key] = message.data.value
}

export const updateAction = ({ state }: Context, message: EndActionMessage) => {
  const app = state.apps[message.appName]
  const action = message.data
  const id = `${action.actionId}_${action.executionId}`

  app.actions[id].isRunning = false
}

export const addMutations = ({ state }: Context, message: MutationsMessage) => {
  const mutations = message.data

  // For backwards compatability, ensure delimiter
  mutations.mutations.forEach((mutation) => {
    mutation.delimiter = mutation.delimiter || '.'
  })

  const id = `${mutations.actionId}_${mutations.executionId}`
  const operator =
    state.apps[message.appName].actions[id].operators[mutations.operatorId]

  operator.events = operator.events.concat(
    mutations.mutations.map((mutation) => ({
      type: EventType.Mutation,
      data: mutation,
    }))
  )

  message.data.mutations.forEach(runMutation(state.apps[message.appName].state))
}

export const updateEffect = ({ state }: Context, message: EffectMessage) => {
  const effect = message.data
  const id = getActionId(effect)
  const action = state.apps[message.appName].actions[id]
  const operator =
    state.apps[message.appName].actions[id].operators[effect.operatorId]
  const existingEvent = operator.events.find(
    (event) =>
      event.type === EventType.Effect &&
      (event.data as Effect).effectId === effect.effectId
  )

  if (existingEvent) {
    Object.assign(existingEvent.data, effect)
  } else {
    operator.events.push({
      type: EventType.Effect,
      data: effect,
    })
  }

  if (effect.error) {
    action.hasError = true
  }
}

export const getMessage = (_, value: Message) =>
  ({
    ...value.message,
    appName: value.appName,
  } as AppMessage<any>)

export const updateOperatorAsync = () => () => {}
