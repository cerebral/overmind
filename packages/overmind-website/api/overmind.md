# Overmind

The **Overmind** class is used to create the application instance. To connect your application to components your create a connect function using the related view package. 

```marksy
h(Example, { name: "api/app_initialize" })
```

You can pass a second argument to the **App** constructor. This is an options object with the following properties:

## addMutationListener

It is possible to listen to all mutations performed in Overmind. This allows you to create special effects based on mutations within a certain domain of your app, or whatever else you come up with.

```marksy
h(Example, { name: "api/app_addMutationListener" })
```


## options.devtools
If you develop your app on localhost the application connects to the devtools on **localhost:3031**. You can change this in case you need to use an IP address, the devtools is configured with a different port or you want to connect to localhost (with default port) even though the app is not on localhost.

```marksy
h(Example, { name: "api/app_options_devtools" })
```

## options.logProxies
By default, in **development**, Overmind will make sure that any usage of **console.log** will log out the actual object/array, instead of any proxy wrapping it. This is most likely what you want. If you want to turn off this behaviour, set this option to **true**.

```marksy
h(Example, { name: "api/app_options_logproxies" })
```

## options.name
If you have multiple instances of Overmind on the same page you can name your app to differentiate it in the devtools.

```marksy
h(Example, { name: "api/app_options_name" })
```