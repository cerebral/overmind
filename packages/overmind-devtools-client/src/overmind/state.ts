import { derived } from 'overmind'
import {
  Action,
  App,
  Apps,
  Component,
  Flush,
  GroupedComponents,
  Tab,
  HistoryRecord,
  ExecutionType,
  MutationsMessage,
  HistoryRecordType,
  MutationHistoryRecord,
  FlushMessage,
  FlushHistoryRecord,
  EffectMessage,
  EffectHistoryRecord,
  OperatorsByPath,
  Chart,
  ActionsListItemType,
} from './types'

type State = {
  isConnecting: boolean
  error: string
  port: number
  apps: Apps
  currentAppName: string
  newPortValue: string
  currentTab: Tab
  showApps: boolean
  currentAction: Action
  currentChart: Chart
  currentApp: App
  componentsMounted: Component[]
  componentsUpdateCount: number
  componentsStatePathCount: number
  flushes: Flush[]
  flushesMutationsCount: number
  flushesStatePathCount: number
  groupedComponents: GroupedComponents
  history: HistoryRecord[]
  hasActionsError: boolean
  actionsCount: number
  currentOperatorsByPath: OperatorsByPath[]
  isShowingRuntime: boolean
  actionsSplitSize: number
  chartsSplitSize: number
  isExecutingAction: boolean
}

const state: State = {
  isShowingRuntime: false,
  isConnecting: true,
  error: null,
  showApps: false,
  currentAppName: null,
  actionsSplitSize: 200,
  chartsSplitSize: 200,
  port: 3031,
  apps: {},
  newPortValue: '',
  currentTab: Tab.State,
  isExecutingAction: false,
  currentApp: derived((state: State) => {
    return state.apps[state.currentAppName]
  }),
  actionsCount: derived((state: State) => {
    if (!state.currentApp) return 0
    return state.currentApp.actionsList.reduce(
      (aggr, item) =>
        aggr +
        (item.type === ActionsListItemType.ACTION ? 1 : item.actionIds.length),
      0
    )
  }),
  hasActionsError: derived((state: State) => {
    return Boolean(
      state.currentApp &&
        Object.values(state.currentApp.actions).some(
          (action) => action.hasError
        )
    )
  }),
  componentsMounted: derived((state: State) =>
    Object.keys(state.currentApp.components).reduce((aggr, key) => {
      if (state.currentApp.components[key].isMounted) {
        return aggr.concat(state.currentApp.components[key])
      }

      return aggr
    }, [] as Component[])
  ),
  componentsUpdateCount: derived((state: State) =>
    state.componentsMounted.reduce(
      (aggr, component) => aggr + component.updateCount,
      0
    )
  ),
  componentsStatePathCount: derived((state: State) =>
    state.componentsMounted.reduce(
      (aggr, component) => aggr + component.paths.length,
      0
    )
  ),
  flushes: derived((state: State) =>
    Object.keys(state.currentApp.flushes)
      .sort(
        (idA, idB) =>
          state.currentApp.flushes[idB].flushId -
          state.currentApp.flushes[idA].flushId
      )
      .map((id) => state.currentApp.flushes[id])
  ),
  flushesMutationsCount: derived((state: State) =>
    state.flushes.reduce((aggr, flush) => aggr + flush.mutations.length, 0)
  ),
  flushesStatePathCount: derived(
    (state: State) =>
      state.flushes.reduce((aggr, flush) => {
        return flush.mutations.reduce(
          (aggr, mutation) =>
            aggr.includes(mutation.path) ? aggr : aggr.concat(mutation.path),
          aggr
        )
      }, []).length
  ),
  currentAction: derived(
    (state: State) => state.currentApp.actions[state.currentApp.currentActionId]
  ),
  currentChart: derived(
    (state: State) => state.currentApp.charts[state.currentApp.currentChartId]
  ),
  groupedComponents: derived((state: State) => {
    const components = state.componentsMounted

    return components.reduce((aggr, component) => {
      if (aggr[component.name]) {
        aggr[component.name].components.push(component)
      } else {
        aggr[component.name] = {
          name: component.name,
          components: [component],
          isCollapsed: !state.currentApp.expandedComponents.includes(
            component.name
          ),
        }
      }

      return aggr
    }, {} as GroupedComponents)
  }),
  history: derived((state: State) => {
    return state.currentApp.messages.reduce((aggr, message) => {
      switch (message.type) {
        case ExecutionType.MUTATIONS: {
          const mutationsMessage = message as MutationsMessage

          return aggr.concat(
            mutationsMessage.data.mutations.map((mutation) => {
              const mutationRecord: MutationHistoryRecord = {
                type: HistoryRecordType.Mutation,
                data: mutation,
                actionName: mutationsMessage.data.actionName,
              }

              return mutationRecord
            })
          )
        }
        case ExecutionType.FLUSH: {
          const flushMessage = message as FlushMessage
          const flush = state.currentApp.flushes[flushMessage.data.flushId]
          const flushRecord: FlushHistoryRecord = {
            type: HistoryRecordType.Flush,
            data: {
              ...flushMessage.data,
              components: flush.components.map(
                (componentId) => state.currentApp.components[componentId].name
              ),
              derived: flush.derived.map(
                (derivedId) => state.currentApp.derived[derivedId]
              ),
              isCollapsed: true,
            },
            actionName: flushMessage.data.actionName,
          }

          return aggr.concat(flushRecord)
        }
        case ExecutionType.EFFECT: {
          const effectMessage = message as EffectMessage
          const effectRecord: EffectHistoryRecord = {
            type: HistoryRecordType.Effect,
            data: effectMessage.data,
            actionName: effectMessage.data.actionName,
          }

          return aggr.concat(effectRecord)
        }
      }

      return aggr
    }, [])
  }),
  currentOperatorsByPath: derived((state: State) => {
    const operators = Object.keys(state.currentAction.operators)
      .map((id) => Number(id))
      .sort((a, b) => a - b)
      .map((operatorId) => state.currentAction.operators[operatorId])

    return operators.reduce((aggr, operator): OperatorsByPath[] => {
      let currentValue = state.currentAction.value
      const traversePath = operator.path.slice()
      traversePath.unshift('')
      // eslint-disable-next-line
      traversePath.reduce((childrenByPath, key, index) => {
        const isLastKey = index === traversePath.length - 1
        const matchingChildren = childrenByPath.find(
          (children) => children[0].path === key
        )
        const lastChildByPath = matchingChildren
          ? matchingChildren[matchingChildren.length - 1]
          : null

        if (isLastKey) {
          const newChild = {
            path: key,
            operator,
            childrenByPath: [],
            value: matchingChildren
              ? lastChildByPath.operator.result
              : currentValue,
          }

          matchingChildren
            ? matchingChildren.push(newChild)
            : childrenByPath.push([newChild])

          // eslint-disable-next-line
          return
        }

        currentValue = lastChildByPath.value

        return lastChildByPath.childrenByPath
      }, aggr)

      return aggr
    }, [])
  }),
}

export default state
