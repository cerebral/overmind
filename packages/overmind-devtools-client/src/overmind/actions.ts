import { pipe, wait, fork, Overmind } from 'overmind'
import { Context } from './'

import * as o from './operators'
import {
  ActionsListItemType,
  Component,
  ExecutionType,
  AppMessageType,
  Tab,
} from './types'
import { createApp, isValidJson } from './utils'

export const onInitializeOvermind = async (
  { state, effects }: Context,
  app: Overmind<Context>
) => {
  state.port = effects.utils.getPort()

  effects.connector.onMessage(app.actions.onMessage)
  effects.connector.onDisconnect(app.actions.onDisconnect)

  await effects.connector.connect(state.port)

  state.actionsSplitSize =
    (await effects.storage.get('devtool.actionsSplitSize')) || 200
}

export const onMessage = pipe(
  o.ensureCurrentApp,
  o.ensureApp,
  o.addClientMessage,
  o.getMessage,
  fork('type', {
    [ExecutionType.INIT]: o.addStateAndActions,
    [ExecutionType.RE_INIT]: o.addStateAndActions,
    [ExecutionType.FLUSH]: o.addFlush,
    [ExecutionType.DERIVED]: o.updateDerived,
    [ExecutionType.MUTATIONS]: o.addMutations,
    [ExecutionType.EFFECT]: o.updateEffect,
    [ExecutionType.STATE]: o.updateState,
    [ExecutionType.COMPONENT_ADD]: o.addComponent,
    [ExecutionType.COMPONENT_UPDATE]: o.updateComponent,
    [ExecutionType.COMPONENT_REMOVE]: o.removeComponent,
    [ExecutionType.DERIVED_DIRTY]: o.updateFlushWithDerived,
    [ExecutionType.ACTION_START]: o.addAction,
    [ExecutionType.OPERATOR_START]: o.addOperator,
    [ExecutionType.OPERATOR_END]: o.updateOperator,
    [ExecutionType.ACTION_END]: o.updateAction,
    [ExecutionType.OPERATOR_ASYNC]: o.updateOperatorAsync,
    [ExecutionType.GETTER]: o.runGetterMutation,
    [ExecutionType.CHART]: o.addChart,
    [AppMessageType.PORT_EXISTS]: () => {},
  })
)

export const setError = ({ state }: Context, error: string) => {
  state.error = error
}

export const onDisconnect = ({ state }: Context, appName: string) => {
  if (appName in state.apps) {
    state.apps[appName] = createApp({
      name: appName,
    })
    state.apps[appName].connectionState = 'pending'
  }
}

export const changeNewPortValue = ({ state }: Context, port: string) => {
  state.newPortValue = String(Number(port))
}

export const changeTab = ({ state, effects }: Context, tab: Tab) => {
  state.currentTab = tab
  effects.storage.set(`${state.currentApp.name}.currentTab`, tab)
}

export const toggleExpandState = ({ state }: Context, path: string[]) => {
  const pathString = path.join(state.currentApp.delimiter)

  if (state.currentApp.expandedStatePaths.indexOf(pathString) >= 0) {
    state.currentApp.expandedStatePaths.splice(
      state.currentApp.expandedStatePaths.indexOf(pathString),
      1
    )
  } else {
    state.currentApp.expandedStatePaths =
      state.currentApp.expandedStatePaths.concat(pathString)
  }
}

export const selectAction = ({ state }: Context, actionId: string) => {
  for (const index in state.currentApp.actionsList) {
    const item = state.currentApp.actionsList[index]
    if (
      item.type === ActionsListItemType.GROUP &&
      item.id === actionId &&
      state.currentApp.currentActionId === actionId
    ) {
      item.isCollapsed = !item.isCollapsed
      break
    }
  }
  state.currentApp.currentActionId = actionId
}

export const toggleGroupedComponent = ({ state }: Context, name: string) => {
  const index = state.currentApp.expandedComponents.indexOf(name)

  if (index === -1) {
    state.currentApp.expandedComponents.push(name)
  } else {
    state.currentApp.expandedComponents.splice(index, 1)
  }
}

export const selectApp = ({ state }: Context, appName: string) => {
  state.currentAppName = appName
  state.showApps = false
}

export const toggleShowApps = ({ state }: Context) => {
  state.showApps = !state.showApps
}

export const toggleCollapsedComponent = (_: Context, component: Component) => {
  component.isCollapsed = !component.isCollapsed
}

export const toggleQueryingAction = ({ state }: Context) => {
  state.currentApp.isQueryingAction = !state.currentApp.isQueryingAction
}

export const changeActionQuery = ({ state }: Context, query: string) => {
  state.currentApp.actionQuery = query

  const hit = state.currentApp.actionPaths.find((path) =>
    path.startsWith(query)
  )

  if (query.length && hit) {
    state.currentApp.actionQuerySuggestion = hit
  } else {
    state.currentApp.actionQuerySuggestion = ''
  }
}

export const selectQueryAction = (
  { state, effects, actions }: Context,
  path: string
) => {
  const existingActionQuery = state.currentApp.selectedActionQuery

  state.currentApp.selectedActionQuery = path
  state.currentApp.isQueryingAction = false
  state.currentApp.actionQueryPayload = ''
  effects.storage.set(`${state.currentApp.name}.selectedActionQuery`, path)

  if (existingActionQuery === path) {
    actions.executeAction()
  }
}

export const submitQueryAction = ({ state, effects }: Context) => {
  if (!state.currentApp.actionQuerySuggestion) {
    return
  }

  state.currentApp.selectedActionQuery = state.currentApp.actionQuerySuggestion
  state.currentApp.isQueryingAction = false
  state.currentApp.actionQueryPayload = ''

  effects.storage.set(
    `${state.currentApp.name}.selectedActionQuery`,
    state.currentApp.selectedActionQuery
  )
}

export const executeAction = pipe(
  ({ state, effects }: Context) => {
    state.isExecutingAction = true

    const payload = state.currentApp.actionQueryPayload

    if (payload && !isValidJson(payload)) {
      return
    }

    effects.connector.sendMessage(state.currentApp.name, 'executeAction', {
      name: state.currentApp.selectedActionQuery,
      payload: JSON.stringify(
        // eslint-disable-next-line
        eval(
          `(function () { return ${state.currentApp.actionQueryPayload} })()`
        )
      ),
    })
  },
  wait(500),
  ({ state, effects }: Context) => {
    state.currentApp.actionQueryPayload = ''
    effects.storage.set(`${state.currentApp.name}.actionQueryPayload`, '')
    state.isExecutingAction = false
  }
)

export const setActionQueryPayload = (
  { state, effects }: Context,
  payload: string
) => {
  state.currentApp.actionQueryPayload = payload
  effects.storage.set(`${state.currentApp.name}.actionQueryPayload`, payload)
}

export const setState = ({ state }: Context, path: string[]) => {
  state.currentApp.selectedStatePath = path.join(state.currentApp.delimiter)
}

export const undoSettingState = ({ state }: Context) => {
  state.currentApp.selectedStatePath = null
}

export const submitState = ({ state, effects }: Context, newState: string) => {
  const path = state.currentApp.selectedStatePath.split(
    state.currentApp.delimiter
  )

  effects.connector.sendMessage(state.currentApp.name, 'mutation', {
    path,
    // eslint-disable-next-line
    value: JSON.stringify(eval(`(function () { return ${newState} })()`)),
  })

  state.currentApp.selectedStatePath = null
}

export const toggleRuntimeConfig = ({ state }: Context) => {
  state.isShowingRuntime = !state.isShowingRuntime
}

export const refreshApp = ({ state, effects }: Context) => {
  effects.connector.sendMessage(state.currentAppName, 'refresh')
}

export const setAppDataFromStorage = async (
  { state, effects }: Context,
  {
    appName,
    actions,
  }: {
    appName: string
    actions: string[]
  }
) =>
  Promise.all([
    effects.storage.get<string>(`${appName}.selectedActionQuery`),
    effects.storage.get<string>(`${appName}.actionQueryPayload`),
    effects.storage.get<Tab>(`${appName}.currentTab`),
  ]).then(
    ([selectedActionQuery, actionQueryPayload, tab]: [string, string, Tab]) => {
      const actionQuery =
        selectedActionQuery && actions.includes(selectedActionQuery)
          ? selectedActionQuery
          : ''
      state.apps[appName].selectedActionQuery = actionQuery
      state.apps[appName].actionQueryPayload = actionQuery
        ? actionQueryPayload
        : ''
      state.currentTab = tab || state.currentTab
    }
  )

export const updateActionsSplitSize = pipe(
  async ({ state, effects }: Context, size: number) => {
    state.actionsSplitSize = size

    await effects.storage.set('devtool.actionsSplitSize', size)
  }
)

export const updateChartsSplitSize = pipe(
  async ({ state, effects }: Context, size: number) => {
    state.chartsSplitSize = size

    await effects.storage.set('devtool.chartsSplitSize', size)
  }
)

export const clearActions = ({ state }: Context) => {
  state.currentApp.actions = {}
  state.currentApp.actionsList = []
  state.currentApp.currentActionId = null
}

export const selectChart = ({ state }: Context, id: string) => {
  state.currentApp.currentChartId = id
}
