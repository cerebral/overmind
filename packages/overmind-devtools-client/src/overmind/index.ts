import { IContext } from 'overmind'
import {
  createStateHook,
  createActionsHook,
  createEffectsHook,
  createReactionHook,
} from 'overmind-react'

import * as actions from './actions'
import * as effects from './effects'
import state from './state'

export const config = {
  effects,
  actions,
  state,
}

export type Context = IContext<typeof config>

export const useAppState = createStateHook<Context>()
export const useActions = createActionsHook<Context>()
export const useEffects = createEffectsHook<Context>()
export const useReaction = createReactionHook<Context>()
