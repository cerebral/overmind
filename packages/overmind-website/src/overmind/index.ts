import { Overmind, IConfig } from 'overmind'
import state from './state'
import onInitialize from './onInitialize'
import * as actions from './actions'
import * as effects from './effects'
import { createHook } from 'overmind-react'

const config = {
  onInitialize,
  state,
  actions,
  effects,
}

declare module 'overmind' {
  interface Config extends IConfig<typeof config> {}
}

const app = new Overmind(
  config,
  process.env.NODE_ENV === 'production'
    ? {
        devtools: false,
      }
    : {}
)

export const useOvermind = createHook(app)
