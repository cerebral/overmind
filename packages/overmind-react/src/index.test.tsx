import { Overmind, createOvermindMock, IContext } from 'overmind'
import * as React from 'react'
import { render, act } from '@testing-library/react'

import '@testing-library/jest-dom'

import { Provider, createStateHook } from './'

describe('React', () => {
  test('should allow using hooks', () => {
    let renderCount = 0

    const doThis = ({ state }: Context) => {
      state.foo = 'bar2'
    }
    const state = {
      foo: 'bar',
    }
    const actions = {
      doThis,
    }
    const config = {
      state,
      actions,
    }

    type Context = IContext<{
      state: typeof state
      actions: typeof actions
    }>

    const app = new Overmind(config)

    const useState = createStateHook<Context>()

    const FooComponent: React.FunctionComponent = () => {
      const state = useState()
      renderCount++

      return <h1>{state.foo}</h1>
    }

    const { container } = render(
      <Provider value={app}>
        <FooComponent />
      </Provider>
    )

    expect(renderCount).toBe(1)

    act(() => {
      app.actions.doThis()
    })

    expect(renderCount).toBe(2)
    expect(container).toMatchSnapshot()
  })

  test('should allow using hooks with scoped tracking', () => {
    let renderCount = 0

    const doThis = ({ state }: Context) => {
      state.foo.push({ foo: 'bar2' })
    }
    const doThat = ({ state }: Context) => {
      state.foo[0].foo = 'bar3'
    }
    const state = {
      foo: [{ foo: 'bar' }],
    }
    const actions = {
      doThis,
      doThat,
    }
    const config = {
      state,
      actions,
    }

    type Context = IContext<{
      state: typeof state
      actions: typeof actions
    }>

    const app = new Overmind(config)

    const useState = createStateHook<Context>()

    const FooComponent: React.FunctionComponent = () => {
      const state = useState((state) => state.foo[0])
      renderCount++

      return <h1>{state.foo}</h1>
    }

    const { container } = render(
      <Provider value={app}>
        <FooComponent />
      </Provider>
    )

    expect(renderCount).toBe(1)

    act(() => {
      app.actions.doThis()
    })

    expect(renderCount).toBe(1)

    act(() => {
      app.actions.doThat()
    })
    expect(renderCount).toBe(2)

    // This is not showing the expected result, but logging the rendering does, so must be the
    // library messing it up
    expect(container).toMatchSnapshot()
  })
  test('should allow using mocked Overmind', () => {
    let renderCount = 0

    const doThis = ({ state }: Context) => {
      state.foo = 'bar2'
    }
    const state = {
      foo: 'bar',
    }
    const actions = {
      doThis,
    }
    const config = {
      state,
      actions,
    }

    type Context = IContext<{
      state: typeof state
      actions: typeof actions
    }>
    const useState = createStateHook<Context>()

    const FooComponent: React.FunctionComponent = () => {
      const state = useState()
      renderCount++

      return <h1>{state.foo}</h1>
    }

    const mock = createOvermindMock(config)
    const { container } = render(
      <Provider value={mock}>
        <FooComponent />
      </Provider>
    )

    expect(renderCount).toBe(1)
    expect(container).toMatchSnapshot()
  })
  test('should throw an error without provider', () => {
    expect.assertions(1)

    const useState = createStateHook()

    const FooComponent = () => {
      const state = useState()

      return <h1>{(state as any).foo}</h1>
    }

    expect(() => {
      render(<FooComponent />)
    }).toThrow(Error)
  })
})
