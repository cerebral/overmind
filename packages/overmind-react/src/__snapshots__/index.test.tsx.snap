// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React should allow using hooks 1`] = `
<div>
  <h1>
    bar2
  </h1>
</div>
`;

exports[`React should allow using hooks with scoped tracking 1`] = `
<div>
  <h1>
    bar3
  </h1>
</div>
`;

exports[`React should allow using mocked Overmind 1`] = `
<div>
  <h1>
    bar
  </h1>
</div>
`;
